### ISSUE TITLE
Reproducing toy example

### ISSUE THREAD

**Issue by shenberg** [kam1107/RealnessGAN#9]
Hi, I wanted to replicate your results and started with the toy example.

I cloned the repo, created `./OUTPUT/` and `./LOG/` and then ran `scripts/run_MixGauss.sh`

To plot my results, I opened a notebook, loaded the output file, and used a scatter plot:

```python
import numpy as np
import matplotlib.pyplot as plt
import pickle
from pathlib import Path

path = Path('/path/to/RealnessGAN/')

p = pickle.load(open(path / 'OUTPUT/RealnessGAN-MixGauss-1_Extra/5/extra.pk','rb'))
a = np.array(p)
plt.scatter(a[:,0],a[:,1])
```

And the output was not as expected:

![image](https://user-images.githubusercontent.com/653972/83663637-9b459d00-a5d1-11ea-9587-a9785b0b6eb1.png)

From reading `run_MixGauss.sh`, I understand the setting should correspond to "out=5, Giter=3" from Figure 3 in the paper:

![image](https://user-images.githubusercontent.com/653972/83663987-10b16d80-a5d2-11ea-9274-7c460e512400.png)

Any chance for help figuring out why my results are different?

Tech specs: GTX 1080Ti, Ubuntu 18.04, torch 1.5, nvidia-440 on CUDA 10.2

**Comment by kam1107** [comment]
Hi, the result displayed in the paper is the **converged** result. I assume the result you showed was obtained at iter 500 (if you haven't changed any parameters in run_MixGauss.sh). For this setting (out=5, Giter=3), 500 iter is more than enough, i.e. the model could diverge at 500 iter. If you search earlier checkpoints you might be able to find one that gives similar results as the paper.

**Comment by shenberg** [comment]
That was my thought as well, however:

100 iterations:
![image](https://user-images.githubusercontent.com/653972/83736681-a98bcb80-a65a-11ea-8c20-f125b577eae4.png)

200 iterations:
![image](https://user-images.githubusercontent.com/653972/83736761-c58f6d00-a65a-11ea-8991-70dc0a6727e9.png)

300 iters:
![image](https://user-images.githubusercontent.com/653972/83736992-17d08e00-a65b-11ea-8df4-7ce4c2d2d738.png)

400 iters:
![image](https://user-images.githubusercontent.com/653972/83737085-39317a00-a65b-11ea-9bc2-022fcbf75c44.png)

Unfortunately, none of them look converged (loss was static mostly as well). I increased the amount of outputs by setting `--gen-every 25` to look at more of the middle and the results are the same: no convergence.

**Comment by kam1107** [comment]
I just re-run the experiment and it works fine... could you please confirm you cloned the latest version?

**Comment by kam1107** [comment]
To plot I use:
`import os
from os.path import join
import pickle
import torch
import matplotlib.pyplot as plt
import numpy as np

ft_size=40
tick_size=30
label_size=40

plt.rcParams.update({'font.size': ft_size})
plt.rcParams.update({'figure.autolayout': True})
plt.rcParams['axes.facecolor']='w'
plt.rcParams['grid.color'] = '#C0C0C0'
plt.rc('xtick', labelsize=tick_size)
plt.rc('ytick', labelsize=tick_size)
fig = plt.figure(figsize=(8,8))

cwd = join(os.getcwd(), 'RealnessGAN-MixGauss-1_Extra')
dirs = os.listdir(cwd)


with open(join(cwd, '5', 'extra.pk'), 'rb') as f:
    data = pickle.load(f)
x_axis = []
y_axis = []

for i in range(10000):
    x_axis.append(data[i][0])
    y_axis.append(data[i][1])

plt.scatter(x_axis, y_axis, alpha=1/8)
plt.ylim([-3.5, 3.5])
plt.xlim([-3.5, 3.5])

plt.show()`

and the result looks like this:
![image](https://user-images.githubusercontent.com/13587147/83742042-83caea80-a68b-11ea-9815-1f351161add7.png)

**Comment by shenberg** [comment]
Hi, I re-cloned and re-ran the script, with the same results. I then re-ran the script using the CPU and got the same poor results. (I did this by adding `--cuda False` as a command-line parameter and fixing a tiny bug in `train_toy.py`:
line 51 is missing:
```python
else:
    device = 'cpu'
```
). I'm at a loss, as taking the GPU out of the equation removes the biggest differentiating factor...

**Comment by shenberg** [comment]
I also ran the code on CPU on a macbook pro. Still no convergence...

**Comment by kam1107** [comment]
Which torch version are you using?

**Comment by shenberg** [comment]
`torch==1.5.0`

**Comment by kam1107** [comment]
Just checked I used torch==1.1.0 in experiments. It seems that different torch version has different impact on this toy dataset. I will update README to specify which version should be used and otherwise you might need to re-search parameters. Thanks for reporting this and please let me know if it still doesn't work.

**Comment by shenberg** [comment]
Ok, `torch=1.1.0` and `torchvision==0.3.0` solved the issue. Very very strange that this is the case.

![image](https://user-images.githubusercontent.com/653972/83950191-794a5580-a831-11ea-8a73-3b06ed231df2.png)

**Comment by shenberg** [comment]
Hmm, `torch==1.2.0` breaks already. I thought it was the breaking change of `uint8`->`bool` in masking (`loss.py` lines 80-81 need `*` to be replaced by `&` inside the indexing for `l`, `u`) but this doesn't fix the issue in 1.2.0 (it doesn't break on 1.1.0 so it should be strictly better anyway, but no fix = no PR)

**Comment by shenberg** [comment]
So, out of curiosity, I tried to replicate some more of the figure, and unfortunately nothing really worked when cranking up `num_outcomes`:
out = 10, Giter=6

![image](https://user-images.githubusercontent.com/653972/84432163-1d1e7180-ac35-11ea-97e9-daf262ab48b1.png)

out=20, Giter=14

![image](https://user-images.githubusercontent.com/653972/84432494-8dc58e00-ac35-11ea-8824-42cb179a66ed.png)

Any idea what's going on?
Just to make sure, I checked all five checkpoints, none of them had a good result

**Comment by kam1107** [comment]
Hi, the idea of this toy experiment is to show that more outcomes provide stronger constraints on G. Hence, G needs to increase its update frequency to keep up with D. You can try to increase the total number of iterations e.g. 10000 and inspect how the converged pattern evolve. Usually more outcomes requires more iterations to converge. Just to mention that this repo is refactored from my experimenting code so you might find some results cannot be 100% reproduced with the exactly same setting, but in general similar effect can be achieved.

**Comment by shenberg** [comment]
`num_outcomes=20` and `G_updates=14` (similar to paper out=20,Giter=14) converges to all 9 modes of the distribution in 5000 iterations. Not ideal but ok.

**Comment by shenberg** [comment]
I just wanted to say, thanks for putting in the time and effort for a random person on the internet. You're great!

**Comment by shenberg** [comment]
I found the issue, made the tiny fix necessary and confirmed that a few of the settings from Fig. 2 in the paper are reproduced correctly. [PR in Github](https://github.com/kam1107/RealnessGAN/pull/10)

In the toy example, the anchors are created once at training start, while with the real code (`train.py`) they are re-generated every iteration. I changed the toy example to behave the same way as in `train.py` and things started behaving much much better.
